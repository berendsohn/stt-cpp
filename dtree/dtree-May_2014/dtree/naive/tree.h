// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
