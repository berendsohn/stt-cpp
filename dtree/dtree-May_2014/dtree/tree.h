// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
