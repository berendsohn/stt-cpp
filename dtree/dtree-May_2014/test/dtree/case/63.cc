#define DTREE_CONFIG EndTreeWithDesc<WithEvert<WithEvert<WithAncDescAggr<FreeSemigroup, NullPredicate_, WithAncValue<FreeGroup, WithAncAggr<FreeSemigroup, NullPredicate_, WithAncDescAggr<FreeSemigroup, NullPredicate_, WithDescValueWithDesc<FreeGroup, WithAncDescAggr<dtree::Count<int>, NullPredicate_, Begin > > > > > > > > >
#include "test/dtree/tree.cc"
