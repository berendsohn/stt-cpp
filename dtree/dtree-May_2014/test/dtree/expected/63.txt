<<<
  1: v569,000,703,000 v787,000,f68,000
  2: v7b9,000,556,000 vde9,000,812,000 -> 3
  3: v276,000,f12,000 v51e,000,934,000
  4: v421,000,069,000 veed,000,330,000 -> 14
  5: v68c,000,881,000 va85,000,2d0,000
  6: vaa5,000,430,000 v69a,000,797,000 -> 3
  7: v068,000,608,000 v09c,000,87d,000
  8: vf32,000,2cc,000 ved6,000,f8b,000 -> 3
  9: v84a,000,0ba,000 v3bb,000,19c,000 -> 3
  10: v170,000,852,000 vff4,000,8cb,000
  11: veff,000,bc2,000 vdec,000,3f7,000 -> 8
  12: veda,000,e47,000 v0a2,000,ea3,000 -> 2
  13: vd74,000,08e,000 v645,000,c09,000 -> 14
  14: v157,000,375,000 v088,000,9a9,000
  15: v429,000,8c1,000 vb6c,000,0fa,000 -> 3
>>>
<<<
  @1: a1 o0 v569,000,703,000 a1 o0 v569,000,703,000 a000,000 o000,000 a000,000 v787,000,f68,000 a1 o0 v569,000,703,000 a1 o0 v569,000,703,000 a000,000 o000,000 a000,000 v787,000,f68,000 a000,000 o000,000 f1 f0

  @5: a1 o0 v68c,000,881,000 a1 o0 v68c,000,881,000 a000,000 o000,000 a000,000 va85,000,2d0,000 a1 o0 v68c,000,881,000 a1 o0 v68c,000,881,000 a000,000 o000,000 a000,000 va85,000,2d0,000 a000,000 o000,000 f1 f0

  @7: a1 o0 v068,000,608,000 a1 o0 v068,000,608,000 a000,000 o000,000 a000,000 v09c,000,87d,000 a1 o0 v068,000,608,000 a1 o0 v068,000,608,000 a000,000 o000,000 a000,000 v09c,000,87d,000 a000,000 o000,000 f1 f1

    /3: a1 o5 v276,f32,f12,2cc a1 o5 v276,f32,f12,2cc a000,000 o276,f12 a000,000 v51e,ed6,934,f8b a1 o5 v276,f32,f12,2cc a1 o5 v276,f32,f12,2cc a000,000 o276,f12 a000,000 v51e,ed6,934,f8b a000,000 o000,000 f1 f0
         .'9: a1 o1 v84a,429,0ba,8c1 a1 o1 v84a,429,0ba,8c1 a000,000 o000,000 a000,000 v3bb,000,19c,000 a1 o1 v84a,429,0ba,8c1 a1 o1 v84a,429,0ba,8c1 a000,000 o000,000 a000,000 v3bb,000,19c,000 a000,000 o000,000 f1 f1
       .'15: a1 o3 v429,eda,8c1,e47 a1 o3 v429,eda,8c1,e47 a429,8c1 o429,8c1 a429,8c1 vb6c,000,0fa,000 a1 o3 v429,eda,8c1,e47 a1 o3 v429,eda,8c1,e47 a429,8c1 o429,8c1 a429,8c1 vb6c,000,0fa,000 a000,000 o000,000 f1 f1
         '.6: a1 o1 vaa5,429,430,8c1 a1 o1 vaa5,429,430,8c1 a000,000 o000,000 a000,000 v69a,000,797,000 a1 o1 vaa5,429,430,8c1 a1 o1 vaa5,429,430,8c1 a000,000 o000,000 a000,000 v69a,000,797,000 a000,000 o000,000 f1 f0
     ->12: a2 o5 veda,276,e47,f12 a2 o5 veda,276,e47,f12 aeda,e47 oeda,e47 aeda,e47 v0a2,000,ea3,000 a2 o5 veda,276,e47,f12 a2 o5 veda,276,e47,f12 aeda,e47 oeda,e47 aeda,e47 v0a2,000,ea3,000 a000,000 o000,000 f0 f1
        \2: a1 o0 v7b9,eda,556,e47 a1 o0 v7b9,eda,556,e47 a7b9,556 o000,000 a7b9,556 vde9,0a2,812,ea3 a1 o0 v7b9,eda,556,e47 a1 o0 v7b9,eda,556,e47 a7b9,556 o000,000 a7b9,556 vde9,0a2,812,ea3 ade9,812 o000,000 f1 f1
  @8: a2 o0 vf32,000,2cc,000 a2 o0 vf32,000,2cc,000 af32,2cc o000,000 af32,2cc ved6,000,f8b,000 a2 o0 vf32,000,2cc,000 a2 o0 vf32,000,2cc,000 af32,2cc o000,000 af32,2cc ved6,000,f8b,000 aed6,f8b o000,000 f0 f0
    \11: a1 o0 veff,f32,bc2,2cc a1 o0 veff,f32,bc2,2cc a000,000 o000,000 a000,000 vdec,ed6,3f7,f8b a1 o0 veff,f32,bc2,2cc a1 o0 veff,f32,bc2,2cc a000,000 o000,000 a000,000 vdec,ed6,3f7,f8b a000,000 o000,000 f0 f1

  @10: a1 o0 v170,000,852,000 a1 o0 v170,000,852,000 a000,000 o000,000 a000,000 vff4,000,8cb,000 a1 o0 v170,000,852,000 a1 o0 v170,000,852,000 a000,000 o000,000 a000,000 vff4,000,8cb,000 a000,000 o000,000 f1 f1

  @14: a1 o1 v157,000,375,000 a1 o1 v157,000,375,000 a157,375 o000,000 a157,375 v088,000,9a9,000 a1 o1 v157,000,375,000 a1 o1 v157,000,375,000 a157,375 o000,000 a157,375 v088,000,9a9,000 a088,9a9 o000,000 f1 f1
   ->13: a1 o1 vd74,157,08e,375 a1 o1 vd74,157,08e,375 a000,000 o000,000 a000,000 v645,000,c09,000 a1 o1 vd74,157,08e,375 a1 o1 vd74,157,08e,375 a000,000 o000,000 a000,000 v645,000,c09,000 a000,000 o000,000 f1 f0
    \4: a1 o0 v421,157,069,375 a1 o0 v421,157,069,375 a421,069 o000,000 a421,069 veed,088,330,9a9 a1 o0 v421,157,069,375 a1 o0 v421,157,069,375 a421,069 o000,000 a421,069 veed,088,330,9a9 aeed,330 o000,000 f1 f0
>>>
